### PR TITLE
Add Keras Tuner docs

### PR DESCRIPTION
**Pull request by haifeng-jin**
Adding Keras Tuner docs.

The guides are added to `keras.io/guides/keras-tuner/`.
The API docs are added to `keras.io/api/keras-tuner/`.
The landing page is added to `keras.io/keras-tuner/`.
The getting started for KT is added to `keras.io/keras-tuner/getting_started/`.

### CODE REVIEW

File: guides/keras-tuner/distributed_tuning.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
+"""
+Title: Distributed Tuning on Multiple GPUs and Machines
```
Comment (fchollet) [inline guides/keras-tuner/distributed_tuning.py:2]: Please switch the title to "Distributed hyperparameter tuning", in the 3 files (we need to make it clearly different from "distributed training")

File: guides/keras-tuner/distributed_tuning.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+Author: Tom O'Malley, Haifeng Jin
+Date created: 2019/10/24
+Last modified: 2021/06/02
+Description: Tuning the hyperparameters of the models with multiple GPUs and multiple machines.
+"""
+"""
+
+Keras Tuner makes it easy to perform distributed hyperparameter search. No changes to your code are needed to scale up from running single-threaded locally to running on dozens or hundreds of workers in parallel. Distributed Keras Tuner uses a chief-worker model. The chief runs a service to which the workers report results and query for the hyperparameters to try next. The chief should be run on a single-threaded CPU instance (or alternatively as a separate process on one of the workers).
```
Comment (fchollet) [inline guides/keras-tuner/distributed_tuning.py:10]: Please shorten the lines in the `py` file to be ~80 chars or so, it makes review and source control easier. No need to regenerate the other files, since we don't use them for reviewing or source control

File: guides/keras-tuner/distributed_tuning.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+- All workers should have access to a centralized file system to which they can write their results.
+- All workers should be able to access the necessary training and validation data needed for tuning.
+- To support fault-tolerance, `overwrite` should be kept as `False` in `Tuner.__init__` (`False` is the default).
+
+Example bash script for chief service (sample code for `run_tuning.py` at bottom of page):
+
+"""
+
+"""shell
+!export KERASTUNER_TUNER_ID="chief"
```
Comment (fchollet) [inline guides/keras-tuner/distributed_tuning.py:33]: No ! for shell command lines. Did you check that the generated ipynb runs?

File: guides/keras-tuner/subclass_tuner.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
+"""
+Title: Subclassing Tuner for Custom Training Loops
```
Comment (fchollet) [inline guides/keras-tuner/subclass_tuner.py:2]: We only capitalize class names in titles (see https://keras.io/guides/)

In this case please switch the title to "Writing your own Tuner to support a custom training loop"

File: scripts/guides_master.py [hunk 5]
```diff
@@ -1,3 +1,18 @@
...
+    'path': 'keras-tuner/',
+    'title': 'Hyperparameter Tuning',
+    'toc': True,
+    'children': [
+        {
+            'path': 'distributed_tuning',
+            'title': 'Distributed Tuning on Multiple GPUs and Machines',
+        },
+        {
+            'path': 'subclass_tuner',
```
Comment (fchollet) [inline scripts/guides_master.py:11]: "custom_tuner" will be more searchable as part of the URL

File: scripts/kt_api_master.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+                'kerastuner.HyperModel.build',
+                'kerastuner.applications.HyperXception',
+                'kerastuner.applications.HyperResNet',
+            ]
+        },
+        {
+            'path': 'oracles',
+            'title': 'The Oracle class',
+            'generate': [
+                'kerastuner.oracles.BayesianOptimization',
```
Comment (fchollet) [inline scripts/kt_api_master.py:81]: Each specific Oracle should have its own page.

I would recommend:

"KerasTuner Oracles": a landing page with its own template (similar to this https://keras.io/api/layers/)
Subpages:
- The base Oracle class
- RandomSearch
- BayesianOptimization
- Hyperband

File: scripts/kt_api_master.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+                'kerastuner.Tuner.get_state',
+                'kerastuner.Tuner.load_model',
+                'kerastuner.Tuner.run_trial',
+                'kerastuner.Tuner.save_model',
+                'kerastuner.Tuner.search',
+                'kerastuner.Tuner.set_state',
+            ]
+        },
+        {
+            'path': 'hypermodels',
```
Comment (fchollet) [inline scripts/kt_api_master.py:68]: Let's have the same setup for hypermodels and tuners: a landing page explaning basic concepts and showing basic examples, then a page for the base class, then a page for each specific subclass.

Here is another example of a similar structure: https://keras.io/api/applications/

File: scripts/kt_api_master.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        },
+        {
+            'path': 'oracles',
+            'title': 'The Oracle class',
+            'generate': [
+                'kerastuner.oracles.BayesianOptimization',
+                'kerastuner.oracles.Hyperband',
+                'kerastuner.oracles.RandomSearch',
+                'kerastuner.Oracle',
+                'kerastuner.Oracle._populate_space',
```
Comment (fchollet) [inline scripts/kt_api_master.py:85]: We can't document private methods. Either we hide them, or, if they're important, we make them public in a new release.

File: templates/api/keras-tuner/index.md [hunk 9]
```diff
@@ -0,0 +1,16 @@
+# Models API
```
Comment (fchollet) [inline templates/api/keras-tuner/index.md:1]: This template will need updating, ping me when ready

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! Great job!

File: guides/md/keras-tuner/distributed_tuning.md [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+```
+</div>
+### Data parallelism with tf.distribute
+
+Keras Tuner also supports data parallelism via [tf.distribute](https://www.tensorflow.org/tutorials/distribute/keras). Data parallelism and distributed tuning can be combined. For example, if you have 10 workers with 4 GPUs on each worker, you can run 10 parallel trials with each trial training on 4 GPUs by using [tf.distribute.MirroredStrategy](https://www.tensorflow.org/api_docs/python/tf/distribute/MirroredStrategy). You can also run each trial on TPUs via [tf.distribute.experimental.TPUStrategy](https://www.tensorflow.org/api_docs/python/tf/distribute/experimental/TPUStrategy). Currently [tf.distribute.MultiWorkerMirroredStrategy](https://www.tensorflow.org/api_docs/python/tf/distribute/experimental/MultiWorkerMirroredStrategy) is not supported, but support for this is on the roadmap.
+
+
+### Example code
+
+When the enviroment variables described above are set, the example below will run distributed tuning and will also use data parallelism within each trial via `tf.distribute`. The example loads MNIST from `tensorflow_datasets` and uses hyperband for the hyperparameter search.
```
Comment (8bitmp3) [inline guides/md/keras-tuner/distributed_tuning.md:73]: Suggestions:

- "will... will.." -> "will run.. and use data parallelism..."
- Hyperband (capitalized) + link to https://arxiv.org/pdf/1603.06560.pdf

```suggestion
When the enviroment variables described above are set, the example below will run distributed tuning and will also use data parallelism within each trial via `tf.distribute`. The example loads MNIST from `tensorflow_datasets` and uses [Hyperband](https://arxiv.org/pdf/1603.06560.pdf) for the hyperparameter search.
```

File: guides/md/keras-tuner/distributed_tuning.md [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+    optimizer = hp.Choice("optimizer", ["adam", "sgd"])
+    model.compile(
+        optimizer, loss="sparse_categorical_crossentropy", metrics=["accuracy"]
+    )
+    return model
+
+
+def main():
+    """Runs the hyperparameter search."""
+    tuner = kt.Hyperband(
```
Comment (8bitmp3) [inline guides/md/keras-tuner/distributed_tuning.md:120]: Can you double check if we should call a tuner with `kt.tuners.Hyperband` instead of `kt.Hyperband`?

e.g. `from kerastuner.tuners import Hyperband` or `RandomSearch` (https://keras-team.github.io/keras-tuner/examples/helloworld/)

It seems to work in both instances but the latter may be the correct way? (Will  update this here too https://www.tensorflow.org/tutorials/keras/keras_tuner)

Comment (haifeng-jin) [reply guides/md/keras-tuner/distributed_tuning.md:120]: Yes, the `kt.tuner.Hyperband` is importing it from the source file, which is close to where the implementation is. The suggested way of using it would be `kt.Hyperband`.

@fchollet Do you have any comment on this?

Comment (8bitmp3) [reply guides/md/keras-tuner/distributed_tuning.md:120]: Thanks for the reply @haifeng-jin ! Can we suggest to users that if they want to use built-in tuners, they can use either way?

For example, in this blog post - https://blog.tensorflow.org/2020/01/hyperparameter-tuning-with-keras-tuner.html :

> Keras Tuner comes with Bayesian Optimization, Hyperband, and Random Search algorithms built-in...
>
> ```
> tuner = kt.tuners.BayesianOptimization(...)
> ```
> ...
> ```
> tuner = kt.Hyperband(...)
> ```

Comment (fchollet) [reply guides/md/keras-tuner/distributed_tuning.md:153]: We'd generally recommend using top-level symbols (like `kt.Hyperband`), referring to the submodules is pure cognitive overhead for users.

File: templates/guides/keras-tuner/index.md [hunk 12]
```diff
@@ -0,0 +1,12 @@
+# Hyperparameter Tuning
+
+These guides are for using Keras Tuner to tune the hyperparameters of the Keras models.
+
+Most of our guides are written as Jupyter notebooks and can be run in one click in [Google Colab](https://colab.research.google.com/notebooks/welcome.ipynb),
```
Comment (fchollet) [inline templates/guides/keras-tuner/index.md:5]: You can remove these few lines

File: templates/api/keras-tuner/index.md (hunk 9 above)
Comment (fchollet) [reply templates/api/keras-tuner/index.md:1]: We probably want to put content similar to what's on the KT GitHub README

File: guides/md/keras-tuner/distributed_tuning.md (hunk 11 above)
Comment (8bitmp3) [reply guides/md/keras-tuner/distributed_tuning.md:153]: Thank you @fchollet @haifeng-jin 👍 

@MarkDaoust This also means we can close a PR in /docs that proposes `kt.tuners.HyperBand` since it's not a fix (both `kt.tuners.HyperBand` and `kt.HyperBand` work though).

**Review by fchollet — COMMENTED** [review]
What is the purpose of the move `templates/getting_started/ecosystem.md → templates/ecosystem/index.md`?

Such a move should be accompanied by an update of all links that reference the page, plus the addition of a redirect in the original location. We should generate not do this until it's really necessary.

**Comment by haifeng-jin** [comment]
> What is the purpose of the move `templates/getting_started/ecosystem.md → templates/ecosystem/index.md`?
> 
> Such a move should be accompanied by an update of all links that reference the page, plus the addition of a redirect in the original location. We should generate not do this until it's really necessary.

@fchollet 
It seems adding the Keras Tuner subsite will have to add an entry on the nav table on the left of the page.
To better organize the projects in the ecosystem (Keras Tuner, KerasCV and KerasNLP), I think put them all under keras.io/ecosystem/ would be a good idea. Then it would be a duplication to the original ecosystem page.
So now the questions are:

1. How should these project subsites appear in the nav table?
    * Solution 1: Just do not let the subsites appear on the nav table by modifying autogen.py.
    * Solution 2: Let the subsites entry directly under the root of the nav table.
    * Solution 3: Move them under `keras.io/getting_started/ecosystem/`.
    * Solution 4: Create a new page of `keras.io/ecosystem`.
2. If choosing solution 4, what should we do with the original `getting_started/ecosystem.md` and the new `ecosystem/index.md`?

**Comment by fchollet** [comment]
> Solution 2: Let the subsites entry directly under the root of the nav table.

I'd favor this solution. If the nav table ever gets too long there are some entries we can consolidate, like Contributing + Governance or Getting Starter + Why choose Keras.

File: templates/keras-tuner/index.md [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+You can also checkout other versions in our
+[GitHub repository](https://github.com/keras-team/keras-tuner)
+
+## Usage
+
+Here is a simple example of how to use Keras Tuner.
+Import Keras Tuner and TensorFlow.
+```python
+import kerastuner as kt
+import tensorflow as tf
```
Comment (fchollet) [inline templates/keras-tuner/index.md:30]: Prefer `from tensorflow import keras`, there are no TF-specific symbols here

File: templates/keras-tuner/index.md [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+
+Keras Tuner requires **Python 3.6+** and **TensorFlow 2.0+**.
+
+Install latest release:
+
+```
+pip install -U keras-tuner
+```
+
+You can also checkout other versions in our
```
Comment (fchollet) [inline templates/keras-tuner/index.md:21]: check out

**Review by fchollet — COMMENTED** [review]
Looking good! 👍 I pushed a number of copyedits. After that I think we should be good.

**Review by fchollet — APPROVED** [review]
LGTM, thank you for the great work 👍

**Comment by fchollet** [comment]
Note that we should also set up a redirect from the old location of the KT docs to the new one (after this is published).

**Comment by fchollet** [comment]
Also note that in the API docs section, the rendering is slightly off because the docstrings of KerasTuner don't currently conform to the same conventions as the docstrings of Keras. In particular:

```
# Arguments:
```

Should be:

```
Arguments:
```

**Comment by haifeng-jin** [comment]
> Also note that in the API docs section, the rendering is slightly off because the docstrings of KerasTuner don't currently conform to the same conventions as the docstrings of Keras. In particular:
> 
> ```
> # Arguments:
> ```
> 
> Should be:
> 
> ```
> Arguments:
> ```
@fchollet 
I have tracked this problem with this [issue](https://github.com/keras-team/keras-tuner/issues/537).
I also created a [milestone](https://github.com/keras-team/keras-tuner/milestone/1) to track the things to fix in the next release.

Do you think [this docstring](https://github.com/keras-team/keras/blob/master/keras/layers/convolutional.py#L523-L640) is a good example to follow for the docstring style?
Another question is, what should we do with the [old website](https://keras-team.github.io/keras-tuner/)?

**Comment by fchollet** [comment]
> Do you think this docstring is a good example to follow for the docstring style?

Yes, sure. I fixed the docstring formatting in the public docstrings already, but there are various remaining issues.

> Another question is, what should we do with the old website?

Delete it and replace it with a redirect to the new one (one redirect for each individual page). As soon as the new one is live.

Thank you!
